4.14 OLM 1.0 Tech Preview procedures

michaelryanpeter opened the pull request:
<!--- PR title format: [GH#<gh-issue-id>][BZ#<bz-issue-id>][OCPBUGS#<jira-issue-id>][OSDOCS#<jira-issue-id>]: <short-description-of-the-pr> --->

<!--- If your changes apply to the latest release and/or in-development version of OpenShift, open your PR against the `main` branch.

* For more details about the information requested in this template, see:
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/create_or_edit_content.adoc#submit-PR --->

Version(s): 4.14
<!--- Specify the version or versions of OpenShift your PR applies to. -->

Issues:
* [OSDOCS-6686](https://issues.redhat.com//browse/OSDOCS-6686): Set a target version in OLM v1
* [OSDOCS-6687](https://issues.redhat.com//browse/OSDOCS-6687): Removing extensions with OLM v1
* [OSDOCS-6688](https://issues.redhat.com//browse/OSDOCS-6688): How to install an extension with an extension name
* [OSDOCS-6689](https://issues.redhat.com//browse/OSDOCS-6689): How OLM v1 refers to FBC-formatted catalog images
* [OSDOCS-6690](https://issues.redhat.com//browse/OSDOCS-6690): How catalog references are managed on the cluster
* [OSDOCS-6691](https://issues.redhat.com//browse/OSDOCS-6691): How OLM 1.0 resolves installation

<!--- Add a link to the Bugzilla, Jira, or GitHub issue, if applicable. --->

Link to docs preview: [Installing an Operator from a catalog](https://66036--docspreview.netlify.app/openshift-enterprise/latest/operators/olm_v1/olmv1-installing-an-operator-from-a-catalog)
<!--- Add direct link(s) to the exact page(s) with updated content from the preview build. --->

QE review:
- [ ] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->

Additional information:
<!--- Optional: Include additional context or expand the description here.--->

<!--- After you open your PR, ask for review from the OpenShift docs team:
  For community authors: Tag @openshift/team-documentation in a GitHub comment.--->

michaelryanpeter commented on modules/olmv1-about-target-versions.adoc:
Should this module be placed earlier in the document? Perhaps right after Red Hat catalogs?

Is it confusing to have this concept module in the middle of a procedure?

michaelryanpeter commented on modules/olmv1-about-target-versions.adoc:
```diff
@@ -0,0 +1,116 @@
+
+In Operator Lifecycle Manager 1.0, cluster administrators set the target version of an Operator in the Operator's custom resource (CR).
+
+If the Operator's version is not specified in the CR, then OLM 1.0 defaults to the latest stable version in the catalog as the target version. When updates to the Operator are published to the catalog, the Operator automatically updates to the latest stable version.
```
I have taken liberties in this section. I need help to rephrase this so it is accurate.

joelanford replied:
I would strike `stable`.

michaelryanpeter commented on modules/olmv1-about-target-versions.adoc:
```diff
@@ -0,0 +1,116 @@
+  channel: stable-3.8
+  version: 3.8.12
+----
+
```
I have not written about what happens if you only specify a channel. Is that important to document right now?

joelanford replied:
Oops, I didn't see this, see my above comment.

michaelryanpeter commented on modules/olmv1-about-target-versions.adoc:
```diff
@@ -0,0 +1,116 @@
+  channel: stable-3.9
+  packageName: quay-operator
+  version: 999.99.9
+status:
```
Would it help to use code callouts to explain what the different statuses/conditions mean?

michaelryanpeter commented on modules/olmv1-adding-a-catalog.adoc:
```diff
@@ -0,0 +1,75 @@
+----
+<1> Specify the catalog's image in the `spec.source.image` field.
+
+. Run the following command to add the catalog to your cluster. By default, your catalog is installed in the `openshift-catalogd` namespace.
```
Is this correct? I inferred this based on the command to check the status of catalog pods.

michaelryanpeter commented on modules/olmv1-about-target-versions.adoc:
```diff
@@ -0,0 +1,116 @@
+.Command syntax
+[source,terminal]
+----
+$ oc get package <catalog_name>-<operator_name> -o yaml
```
Should this be `<package_name>`, or is <operator_name>` correct?

michaelryanpeter replied:
Thank you! I will fix this.

michaelryanpeter replied:
addressed.

joelanford commented on modules/olmv1-adding-a-catalog.adoc:
```diff
@@ -0,0 +1,75 @@
+redhat-operators      20s
+----
+
+.. Run the following command to check the status of your catalog pods:
++
+[source,terminal]
+----
+$ oc get pod -n openshift-catalogd
+----
++
+.Example output
+[source,text]
+----
+NAME                                           READY   STATUS      RESTARTS   AGE
+catalogd-controller-manager-5b8844bbc4-ps76j   2/2     Running     0          60m
+redhat-operators                               0/1     Completed   0          47s
+----
```
I would suggest changing this slightly such that it focuses only on the `Catalog` API.

The `Catalog` status has conditions, a resolved source, and a phase that are useful to determine the status of the catalog. Perhaps something that outputs `Catalog.status.phase` instead?

michaelryanpeter replied:
@joelanford Do you know the command to output `Catalog.status.phase`?

michaelryanpeter commented on modules/olmv1-red-hat-catalogs.adoc:
```diff
@@ -0,0 +1,63 @@
+[id="olmv1-red-hat-catalogs_{context}"]
+= Red Hat-provided Operator catalogs in OLM 1.0
+
+Red Hat-provided Operator catalogs have been released in the file-based catalog format since {product-title} 4.11.
```
Can delete this first line

michaelryanpeter commented on modules/olmv1-about-catalogs.adoc:
```diff
@@ -0,0 +1,12 @@
+[id="olmv1-about-catalogs_{context}"]
+= About catalogs in OLM 1.0
+
+Operator Lifecycle Manager (OLM) 1.0 introduces the catalogd component to the OLM suite of microservices. Catalogd is a Kubernetes extension that unpacks file-based catalog content for on-cluster clients. Currently, catalogd unpacks catalog content that is packaged and distributed as container images. 
```
```suggestion
Operator Lifecycle Manager (OLM) 1.0 introduces the catalogd component to the OLM suite of microservices. Catalogd is a Kubernetes extension that unpacks catalog content for on-cluster clients. Currently, catalogd unpacks catalog content that is packaged and distributed as container images. 
```

joelanford commented on modules/olmv1-about-target-versions.adoc:
```diff
@@ -0,0 +1,116 @@
+  name: quay-example
+spec:
+  packageName: quay-operator
+  channel: stable-3.8
```
When a `spec.version` is specified, it isn't necessary to use `channel`. In 4.14, version is always a specific version, so `channel` provides no further filtering/constraining beyond what `spec.version` accomplishes.

awgreene commented on modules/olmv1-deleting-an-operator.adoc:
```diff
@@ -0,0 +1,50 @@
+[id="olmv1-deleting-an-operator_{context}"]
+= Deleting an Operator
+
+You can delete an Operator by deleting the Operator's custom resource (CR).
```
Can we call out that the CRD will be deleted as well?

tmshort commented on modules/olmv1-about-target-versions.adoc:
```diff
@@ -0,0 +1,116 @@
+  packageName: quay-operator
+----
+
+If you specify the Operator's target version in the CR, OLM 1.0 installs the specified version. When the target version is specified in the Operator's CR, OLM 1.0 does not change the target version.
```
No mention that the `version` field supports ranges, Masterminds/semver format, not including the hyphen operation.

tmshort replied:
Apparently, version range is not supported in 4.14; but it is there upstream.

michaelryanpeter reviewed (commented):
Notes from the engineering read through.

michaelryanpeter commented on modules/olmv1-about-target-versions.adoc:
```diff
@@ -0,0 +1,116 @@
+[id="olmv1-about-operator-updates_{context}"]
+= About target versions in OLM 1.0
+
+In Operator Lifecycle Manager 1.0, cluster administrators set the target version of an Operator in the Operator's custom resource (CR).
```
I created a follow up ticket to track this work. It is most likely to land in one of the sprints immediately following 4.14 GA. If I can, I will try to pull it in before GA if I can. https://issues.redhat.com/browse/OSDOCS-8218

michaelryanpeter commented:
@jianzhangbjz or @Xia-Zhao-rh PTAL.

jianzhangbjz commented:
Hi @kuiwang02 , could you help review it? Thanks! 
/assign @kuiwang02

kuiwang02 commented on modules/olmv1-deleting-an-operator.adoc:
```diff
@@ -0,0 +1,50 @@
+$ oc get pod -n quay-operator-system
+----
++
+.Example output
```
@michaelryanpeter there are two commands which have two outputs, but your code only one output.
here is the output I tried.
```console
[root@preserve-olm-env2 olmv1review]# oc get operator.operators.operatorframework.io
No resources found
[root@preserve-olm-env2 olmv1review]# oc get pod -n quay-operator-system
No resources found in quay-operator-system namespace.
[root@preserve-olm-env2 olmv1review]# 
```

actually quay-operator-system ns is also deleted, there is no need `oc get pod -n quay-operator-system`

so, 
1, remove the command `oc get pod -n quay-operator-system`
2, if you want to keep it,  better to update the output per command.
3, change `oc get pod -n quay-operator-system` to `oc get ns quay-operator-system`
```console
[root@preserve-olm-env2 olmv1review]# oc get ns quay-operator-system
Error from server (NotFound): namespaces "quay-operator-system" not found
```

I prefer to 3.

michaelryanpeter replied:
@kuiwang02: PTAL

kuiwang02 replied:
it is ok. thanks

kuiwang02 commented:
/assign @Xia-Zhao-rh

michaelryanpeter commented on modules/olmv1-about-target-versions.adoc:
```diff
@@ -0,0 +1,138 @@
+
+In Operator Lifecycle Manager 1.0, cluster administrators set the target version of an Operator declaratively in the Operator's custom resource (CR).
+
+If the Operator's version is not specified in the CR, then OLM 1.0 defaults to the latest version in the catalog as the target version. When updates to the Operator are published to the catalog, the Operator automatically updates to the latest version.
```
Perhaps it needs to be reworded to "the latest version of the default channel in the catalog"?

Xia-Zhao-rh replied:
AFAIK, OLMv1 does not have a concept of a default channel, if NOT setting `operator.spec.channel`,  the bundle can come from any channel.
https://issues.redhat.com/browse/OCPBUGS-16789

michaelryanpeter replied:
Thank you for sharing that link. Good point.

joelanford replied:
Oh, I think there was a bug with the latest version selection. That we fixed recently (but didn't fix in 4.14). I'll find some context and attach it.

joelanford replied:
https://github.com/operator-framework/operator-controller/pull/420

michaelryanpeter replied:
I see two options:
1. We keep the empty version content and add a condition, like: "For Operators with a single channel, OLM 1.0 installs the latest version channel, if the Operator's version is not specified in the CR." (that seems and hard to follow to me.) We should probably add a known issue for this in the release notes if we go this route.
2. We cut this section and focus on what works as expected: pinning a specified version, updates, and specifying a channel with an empty version. If/when the backport lands, we can update the docs to add this part back in. 

My preference for option two. It is more direct, easier to understand, and easier to maintain.

WDYT, @joelanford and @mwerner2113

mwerner2113 replied:
@michaelryanpeter I like option 2 as well. More straightforward.

michaelryanpeter replied:
I took the liberty of removing the section on leaving the version unspecified. I created a bullet to follow up on this issue after GA in this tracking ticket: https://issues.redhat.com/browse/OSDOCS-8218.

Please let me know if you want to discuss this further. I will open a separate PR with the removed content, and we can discuss it there without holding up this PR.

kevinrizza replied:
This seems reasonable to me.

kevinrizza replied:
This seems reasonable to me.

michaelryanpeter commented on modules/olmv1-red-hat-catalogs.adoc:
```diff
@@ -0,0 +1,61 @@
+= Red Hat-provided Operator catalogs in OLM 1.0
+
+Operator Lifecycle Manager (OLM) 1.0 does not include Red Hat-provided Operator catalogs by default. If you want to add a Red Hat-provided catalog to your cluster, create a custom resource (CR) for the catalog and apply it to the cluster. The following custom resource (CR) examples show how to create a catalog resources for OLM 1.0.
+
```
@Xia-Zhao-rh, in the ticket, https://issues.redhat.com//browse/OSDOCS-6689, only these three were specified. I wasn't sure if this was because Marketplace subscriptions aren't supported yet. I will do some research on this today.

michaelryanpeter replied:
@kevinrizza or @tlwu2013 Are you able to clarify this issue? Should the Marketplace catalog be intentionally left out of the catalog examples? Or was that an oversight in the ticket?

kevinrizza replied:
The redhat-marketplace catalog is built by IBM. I believe it is converted to FBC so it should "work" here, but since redhat has less active control over it as the schema evolves I imagine it would be better to leave it out of "try it out" documentation because changes or fixes if things don't work with v1 will take longer to resolve. We definitely don't test this right now with v1, anyway.

michaelryanpeter commented:
@Xia-Zhao-rh I think I have addressed your comments.
* I removed the section on what happens if you leave the target version unspecified. If the bug fix is backported, I will add the content back in.
* Kevin clarified the issue with the Marketplace catalog.

Would you PTAL?

michaelryanpeter commented:
/label peer-review-needed

jldohmann commented on modules/olmv1-adding-a-catalog.adoc:
```diff
@@ -0,0 +1,110 @@
+  spec:
+    source:
+      image:
+        ref: registry.redhat.io/redhat/redhat-operator-index:v4.14
```
since this block has the attribute sub, did you mean to use `{product-version}` here?
```suggestion
        ref: registry.redhat.io/redhat/redhat-operator-index:v{product-version}
```

jldohmann commented on modules/olmv1-deleting-an-operator.adoc:
```diff
@@ -0,0 +1,60 @@
+
+* Run the following commands to verify that your Operator and its resources were deleted:
+
+** Run the following command to verify the Operator was deleted:
```
```suggestion
** Verify the Operator is deleted by running the following command:
```

jldohmann commented on modules/olmv1-updating-an-operator.adoc:
```diff
@@ -0,0 +1,125 @@
+<1> Update the channel to `stable-3.9`.
+<2> Update the version to `3.9.1`
+
+. Run the following command to apply the update to the cluster:
```
```suggestion
. Apply the update to the cluster by running the following command:
```

jldohmann commented on modules/olmv1-updating-an-operator.adoc:
```diff
@@ -0,0 +1,125 @@
++
+[TIP]
+====
+You can run the following command to patch and apply the changes to your Operator's channel and version from the CLI:
```
```suggestion
You can patch and apply the changes to your Operator's channel and version from the CLI by running the following command:
```

jldohmann commented on modules/olmv1-updating-an-operator.adoc:
```diff
@@ -0,0 +1,125 @@
+
+.Verification
+
+* Run the following command verify that the channel and version updates have been applied:
```
```suggestion
* Verify that the channel and version updates have been applied by running the following command:
```

jldohmann reviewed (commented):
Great work! I left some comments and suggestions below

kuiwang02 commented on modules/olmv1-updating-an-operator.adoc:
```diff
@@ -0,0 +1,123 @@
+$ oc get package redhat-operators-quay-operator -o yaml
+----
+
+. Edit your Operator's CR to update the version to `3.9.1`, as shown in the following example:
```
@michaelryanpeter I have new comments for your updated.
you remove channel information, but this example is following the installation section. for installation section, the example has channel, and but when I update it, the example has no channel. it is a little strange.
I think we should keep the same rule.

and the example output still keep channel information. (refer to https://github.com/openshift/openshift-docs/pull/66036/files#diff-3a4664f5b5ccef6dd8762bcc6bc754af9267fa277045a6f9aeabb7807fb32087R102)

so,
1, if you want to remove channel information, please update the output and update the installation section as well.
or
2, if you move channel information back here, no need to change others.

kuiwang02 replied:
if you still want to remove channel, here is the output.
```console
[root@preserve-olm-env2 olmv1review]# oc get operator.operators.operatorframework.io/quay-example -o yaml
apiVersion: operators.operatorframework.io/v1alpha1
kind: Operator
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"operators.operatorframework.io/v1alpha1","kind":"Operator","metadata":{"annotations":{},"name":"quay-example"},"spec":{"packageName":"quay-operator","version":"3.9.1"}}
  creationTimestamp: "2023-10-19T02:39:05Z"
  generation: 2
  name: quay-example
  resourceVersion: "138637"
  uid: 480be93d-e325-4831-ac4a-8bfd451ef3bd
spec:
  packageName: quay-operator
  version: 3.9.1
status:
  conditions:
  - lastTransitionTime: "2023-10-19T02:39:05Z"
    message: resolved to "registry.redhat.io/quay/quay-operator-bundle@sha256:4864bc0d5c18a84a5f19e5e664b58d3133a2ac2a309c6b5659ab553f33214b09"
    observedGeneration: 2
    reason: Success
    status: "True"
    type: Resolved
  - lastTransitionTime: "2023-10-19T02:39:27Z"
    message: installed from "registry.redhat.io/quay/quay-operator-bundle@sha256:4864bc0d5c18a84a5f19e5e664b58d3133a2ac2a309c6b5659ab553f33214b09"
    observedGeneration: 2
    reason: Success
    status: "True"
    type: Installed
  installedBundleResource: registry.redhat.io/quay/quay-operator-bundle@sha256:4864bc0d5c18a84a5f19e5e664b58d3133a2ac2a309c6b5659ab553f33214b09
  resolvedBundleResource: registry.redhat.io/quay/quay-operator-bundle@sha256:4864bc0d5c18a84a5f19e5e664b58d3133a2ac2a309c6b5659ab553f33214b09
```

michaelryanpeter replied:
Oops. I thought I cleaned up the output. Thank you. I will fix it.

I removed the channel information because engineering mentioned that there is no point in specifying both a channel and a version at the same time.

Xia-Zhao-rh commented:
> @Xia-Zhao-rh I think I have addressed your comments.
> 
> * I removed the section on what happens if you leave the target version unspecified. If the bug fix is backported, I will add the content back in.
> * Kevin clarified the issue with the Marketplace catalog.
> 
> Would you PTAL?

Hi, @michaelryanpeter  LGTM, Thanks.

kuiwang02 commented on modules/olmv1-updating-an-operator.adoc:
```diff
@@ -0,0 +1,123 @@
+
+[source,terminal]
+----
+$ oc patch operator.operators.operatorframework.io/quay-example -p \
```
@michaelryanpeter following the installation, if you update it with only updating version (not removing the channel), the command will does not work
```console
[root@preserve-olm-env2 olmv1review]# oc patch operator.operators.operatorframework.io/quay-example -p \
>   '{"spec":{"version":"3.9.1"}}' \
>   --type=merge
operator.operators.operatorframework.io/quay-example patched
[root@preserve-olm-env2 olmv1review]# oc get operator.operators.operatorframework.io/quay-example -o yaml
apiVersion: operators.operatorframework.io/v1alpha1
kind: Operator
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"operators.operatorframework.io/v1alpha1","kind":"Operator","metadata":{"annotations":{},"name":"quay-example"},"spec":{"channel":"stable-3.8","packageName":"quay-operator","version":"3.8.12"}}
  creationTimestamp: "2023-10-19T02:42:21Z"
  generation: 2
  name: quay-example
  resourceVersion: "140524"
  uid: 5b3895d4-e008-42ee-8e45-84ce3c31bcad
spec:
  channel: stable-3.8
  packageName: quay-operator
  version: 3.9.1
status:
  conditions:
  - lastTransitionTime: "2023-10-19T02:43:08Z"
    message: package 'quay-operator' at version '3.9.1' in channel 'stable-3.8' not
      found
    observedGeneration: 2
    reason: ResolutionFailed
    status: "False"
    type: Resolved
  - lastTransitionTime: "2023-10-19T02:43:08Z"
    message: installation has not been attempted as resolution failed
    observedGeneration: 2
    reason: InstallationStatusUnknown
    status: Unknown
    type: Installed
```

so, for such update, you have to set channel because the installation section has channel.
so, I suggest to keep channel. if you want to remove channel still, please remove this patch method.

by the way, if you apply yam file which only have version, no such issue because the channel is removed.

michaelryanpeter replied:
@kuiwang02 I have updated the commands and example outputs to use only the target version. Thank you for pointing out my error.

kevinrizza commented:
/lgtm

kuiwang02 commented:
/label qe-approved

kuiwang02 commented:
/lgtm

Xia-Zhao-rh commented:
/label qe-approved

Xia-Zhao-rh commented:
/lgtm

michaelryanpeter commented on modules/olmv1-red-hat-catalogs.adoc:
```diff
@@ -0,0 +1,61 @@
+    type: image
+    image:
+      ref: registry.redhat.io/redhat/certified-operator-index:v{product-version}
+----
```
@jianzhangbjz PTAL at @kevinrizza's comment here: https://github.com/openshift/openshift-docs/pull/66036#discussion_r1362706304

Engineering's preference is to leave off the Marketplace catalog, since it is built by IBM and we have less control if something doesn't work with OLM 1.0 at this stage of development.

If you feel strongly that this should be included, I am happy to open another PR and we can debate that issue there.

michaelryanpeter replied:
I created issue OSDOCS-8299. After this PR merges, I will create a follow up PR with the Marketplace catalog, so that we can get the change published before Docs Merge Freeze on 10/27.

michaelryanpeter replied:
Follow up Marketplace catalog PR: https://github.com/openshift/openshift-docs/pull/66647

jianzhangbjz replied:
ok, thanks!

michaelryanpeter commented:
/cherrypick enterprise-4.14
